src/Microsoft.DocAsCode.Build.SchemaDriven/Validators/SchemaValidateService.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.SchemaDriven
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.Exceptions;
+
+    using Newtonsoft.Json.Linq;
+    using Newtonsoft.Json.Schema;
+
+    public class SchemaValidateService
+    {
+        private static ConcurrentDictionary<string, SchemaValidateService> _cache = new ConcurrentDictionary<string, SchemaValidateService>();
+
+        private bool _schemaValidationEnabled = true;
+        private object locker = new object();
+
+        private SchemaValidateService(string license = null)
+        {
+            if (!string.IsNullOrEmpty(license))
+            {
+                RegisterLicense(license);
+            }
+        }
+
+        public static SchemaValidateService GetInstance(string license = null)
+        {
+            return _cache.GetOrAdd(license ?? string.Empty, s => new SchemaValidateService(s));
+        }
+
+        public void Validate(object obj, JSchema schema)
+        {
+            if (!_schemaValidationEnabled)
+            {
+                return;
+            }
+
+            var errors = new List<string>();
+            try
+            {
+                ValidateObject(obj, schema, (sender, args) => errors.Add(args.Message));
+            }
+            catch (Exception e)
+            {
+                if (_schemaValidationEnabled)
+                {
+                    lock (locker)
+                    {
+                        if (_schemaValidationEnabled)
```

Comment:
>_schemaValidationEnabled [](start = 28, length = 24)

double checked?

Reply:
https://en.wikipedia.org/wiki/Double-checked_locking

Reply:
[https://en.wikipedia.org/wiki/Double-checked_locking](https://en.wikipedia.org/wiki/Double-checked_locking)

---
In reply to: [152904504](https://github.com/dotnet/docfx/pull/2228#discussion_r152904504) [](ancestors = 152904504)